trie/node.go

```diff
@@ -120,24 +141,25 @@ func decodeShort(hash, buf, elems []byte) (node, error) {
 	if err != nil {
 		return nil, err
 	}
+	flag := nodeFlag{hash: hash}
 	key := compactDecode(kbuf)
 	if key[len(key)-1] == 16 {
 		// value node
 		val, _, err := rlp.SplitString(rest)
 		if err != nil {
 			return nil, fmt.Errorf("invalid value node: %v", err)
 		}
-		return shortNode{key, valueNode(val), hash, false}, nil
+		return &shortNode{key, append(valueNode{}, val...), flag}, nil
```

Comment:
I'm not exactly sure how Go behaves in this case. Are we sure it won't allocate more memory that needed if we `append`? In theory appending elements grows the underlying array exponentially. Maybe appending to an empty array could be a special case, but I can equally easily imagine that it will leave a lot of extra allocated but unused bytes at the end.